lib/speech/speech_recognizer.ts

```diff
@@ -0,0 +1,233 @@
+        }
+        //console.log('Sending chunk of length ' + chunk.length);
+        if (this._connection && this._connection.readyState === 1) {
+            // OPEN
+            if (this._vad && chunk.length === 320) {
+                if (this._vad.process(chunk))
+                    this.consecutiveSilence = 0;
+                else
+                    this.consecutiveSilence++;
+            }
+            if (this.consecutiveSilence === 96) {
+                this._endDetected = true;
+                console.log("VAD threshold reached", this.consecutiveSilence);
+                this._connection.send(undefined, {}, (err ?: Error) => callback(err));
+            } else {
+                this._connection.send(chunk, { binary: true }, (err ?: Error) => callback(err));
+            }
+        } else {
+            this._bufferedMessages.push(chunk);
+            callback();
+        }
+    }
+}
+
+export interface SpeechRecognizerOptions {
+    locale ?: string;
+    nlUrl ?: string;
+    vad ?: Tp.Capabilities.VadApi|null;
+}
+
+export default class SpeechRecognizer extends events.EventEmitter {
+    private _language : string;
+    private _baseUrl : string;
+    private _connection : WebSocket|null;
+    private _vad : Tp.Capabilities.VadApi|null;
+
+    constructor(options : SpeechRecognizerOptions = {}) {
+        super();
+        this._language = options.locale || 'en-US';
+        this._baseUrl = options.nlUrl || URL;
+        this._connection = null;
+        this._vad = options.vad || null;
+    }
+
+    close() {
+        if (!this._connection)
+            return;
+        this._connection.close();
+        this._connection = null;
+    }
+
+    _doConnect() {
+        let url = this._baseUrl + '/' + this._language + '/voice/stream';
+        let connection = new WebSocket(url, {
+            perMessageDeflate: true
+        });
+        return new Promise((callback, errback) => {
+            connection.on('unexpected-response', (req, res) => {
+                errback(new Error(res.statusMessage));
+            });
+            connection.on('open', () => {
+                this._connection = connection;
+                console.log('STT connection opened');
+                this._connection.send(JSON.stringify({ ver: 1 }));
+                callback(connection);
+            });
+            connection.on('close', (code, reason) => {
+                if (code !== 1000) // 1000 = normal closure (eg timeout, or we closed on our side)
+                    console.log('Connection to STT service closed: ' + code + ' ' + reason);
+                this._connection = null;
+            });
+            connection.on('error', (e : Error) => {
+                this._connection = null;
+                this.emit('error', e);
+            });
+        });
+    }
+
+    _ensureConnection() {
+        if (this._connection)
+            return Promise.resolve(this._connection);
+        else
+            return this._doConnect();
```

Comment:
Connections are not reusable in the new protocol, right? So this logic is not correct here.